mypy_ls/plugin.py

```diff
@@ -111,35 +118,77 @@ def pyls_lint(config: Config, workspace: Workspace, document: Document,
 
     """
     settings = config.plugin_settings('mypy-ls')
+    log.info(
+        "lint settings = %s document.path = %s is_saved = %s",
+        settings,
+        document.path,
+        is_saved
+    )
+
     live_mode = settings.get('live_mode', True)
-    args = ['--incremental',
-            '--show-column-numbers',
-            '--follow-imports', 'silent']
+    dmypy = settings.get("dmypy", False)
+
+    if dmypy and live_mode:
+        # dmypy can only be efficiently run on files that have been saved, see:
+        # https://github.com/python/mypy/issues/9309
+        log.warning("live_mode is not supported with dmypy, disabling")
+        live_mode = False
+
+    args = ['--show-column-numbers']
 
     global tmpFile
     if live_mode and not is_saved and tmpFile:
+        log.info("live_mode tmpFile = %s", live_mode)
         tmpFile = open(tmpFile.name, "w")
         tmpFile.write(document.source)
         tmpFile.close()
         args.extend(['--shadow-file', document.path, tmpFile.name])
-    elif not is_saved:
-        return []
+    elif not is_saved and document.path in last_diagnostics:
+        # On-launch the document isn't marked as saved, so fall through and run
+        # the diagnostics anyway even if the file contents may be out of date.
+        log.info(
+            "non-live, returning cached diagnostics len(cached) = %s",
+            last_diagnostics[document.path]
+        )
+        return last_diagnostics[document.path]
 
     if mypyConfigFile:
         args.append('--config-file')
         args.append(mypyConfigFile)
+
     args.append(document.path)
+
     if settings.get('strict', False):
         args.append('--strict')
 
-    report, errors, _ = mypy_api.run(args)
+    if not dmypy:
+        args.extend([
+            "--incremental",
+            "--follow-imports",
+            "silent"
+        ])
+
+        log.info(f"executing mypy args={args}")
+        report, errors, _ = mypy_api.run(args)
+    else:
+        args = ["run", "--"] + args
+
+        log.info(f"executing dmypy args={args}")
+        report, errors, _ = mypy_api.run_dmypy(args)
+
+    log.debug("report: \n" + report)
```

Comment:
Don't interpolate strings in logging statements, if there is an error in the interpolation that will break the program. Instead let the logger handle the interpolation:

```python
log.debug("report:\n%s", report)
```

Please review the other logging statements too.

Reply:
```suggestion
    log.debug("report:\n%s", report)
```